src/SharpLearning.Backend.TensorFlow.Test/TensorFlowRawMnistTest.cs

```diff
@@ -17,18 +17,144 @@ public class TensorFlowRawMnistTest
     {
         const string DownloadPath = "MnistTest";
 
-        //[TestMethod]
+        [TestMethod]
         public void MnistTest()
         {
-            var mnist = Mnist.Load(DownloadPath);
 
-            var trainReader = mnist.GetTrainReader();
-            var batchSize = 64;
-            var (inputBatch, labelBatch) = trainReader.NextBatch(batchSize);
+            const int featureCount = 28 * 28;
+            Assert.AreEqual(784, featureCount);
+            const int classCount = 10;
 
+            using (var g = new TFGraph())
+            {
+                var x = g.Placeholder(TFDataType.Float, new TFShape(-1, featureCount));
+                // Only why to simply set zeros??
+                var W = //g.ZerosLike( // Does not work with ApplyGradientDescent, how do we inialize??
+                    g.VariableV2(new TFShape(featureCount, classCount), TFDataType.Float)
+                    //)
+                    ;
+                var W_zero = g.Const(new float[featureCount, classCount]);
+                g.Assign(W, W_zero);
+
+                var b = //g.ZerosLike(
+                    g.VariableV2(new TFShape(classCount), TFDataType.Float)
+                    //)
+                    ;
+                var b_zero = g.Const(new float[classCount]);
+                g.Assign(b, b_zero);
+
+                var m = g.MatMul(x, W);
+                var y = g.Add(m, b);
+
+                var expectedY = g.Placeholder(TFDataType.Float, new TFShape(-1, classCount));
+
+                var (loss, backprop) = g.SoftmaxCrossEntropyWithLogits(y, expectedY);
+                var crossEntropy = g.ReduceMean(loss);
+
+                var learningRate = g.Const(new TFTensor(0.5f));
+                // Need to do this in loop, can't do it before we have actual variables...
+                // TODO: Is this how to do it???
+                // https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/apply-gradient-descent
+                var adjustedW = g.ApplyGradientDescent(W, learningRate, backprop);
```

Comment:
@nietras The tensorflow way, in python, to apply gradient descent is to use the optimizer classes. They are defined here in python: 
[Optimizer](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/optimizer.py), This is the base class.
[GradientDescent](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/gradient_descent.py), the most basic of the concrete optimizer implementations.

As you can see in the GradientDescent implementation, that optimizer uses the `apply_gradient_descent` operator. So this must be the correct operator. The base optimizer contains all the setup steps, where all trainable variables are extracted and gradients calculated before the actual call to ApplyGradientDescent, which takes care of updating the variables, representing the weights of the network.

I have an ongoing effort started here to implement a simple version of gradient descent in TensorFlowSharp:
(GradientDescent)[https://github.com/mdabros/TensorFlowSharp/blob/gradient-descent-optimizer/TensorFlowSharp/Training/SGD.cs]

We will probably need a high level concept like this in TensorFlowSharp, in order to match the optimizer/trainer type in CNTK. This again depends on how we would like the actual learning/training api to look.